### PR TITLE
Push paint points into a Suspense cache

### PR DESCRIPTION
**Pull request by jcmorrow**
So, this took me a while, mostly not because of Suspense things, but because our graphics code is complicated. I like where I got to, even though there are still some global access patterns that feel a little awkward. Couple of things I have not had time for:

- Doing full playwright tests, rather than just unit tests.
- Putting this together with #7721 , which was the original goal. 
 
TBH, I am kind of swamped with backend things right now, and I don't really feel like I have the bandwidth to keep carrying these front-end PR's along. If nobody picks them up, that's ok, but they might sit here for a little while until I can get back to them.

### CODE REVIEW

File: packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import { TimeStampedPoint } from "@replayio/protocol";
+import { PaintPoint } from "@replayio/protocol";
```
Comment (jcmorrow) [inline packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts:2]: Combine these imports

File: packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import { insertEntrySorted, nextEntry } from "protocol/utils";
+import { mostRecentEntry } from "protocol/utils";
```
Comment (jcmorrow) [inline packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts:4]: combine these imports

File: packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import { ReplayClientInterface } from "shared/client/types";
+import { createWakeable } from "../utils/suspense";
+import { preCacheExecutionPointForTime } from "./PointsCache";
+import { STATUS_RESOLVED, Wakeable } from "./types";
```
Comment (jcmorrow) [inline packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts:8]: Let's separate relative imports

File: packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  wakeable: Wakeable<TimeStampedPointWithPaintHash>;
+}
+
+export enum Status {
+  Cancelled,
+  Failed,
+  Fetching,
+  Finished,
+  NotStarted,
+}
```
Comment (jcmorrow) [inline packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts:25]: Hmmmm... should these things go in a `types` file?

File: packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      wakeable: createWakeable<TimeStampedPointWithPaintHash>(),
+      time,
+    };
+
+    insertEntrySorted(paintPointRequests, request);
+  }
+  throw request.wakeable;
+}
+
+export function imperitavelyGetPaintPointForTime(time: number) {
```
Comment (jcmorrow) [inline packages/bvaughn-architecture-demo/src/suspense/PaintsCache.ts:103]: I do not like how much these get used outside of this file.

File: packages/protocol/analysisManager.ts [hunk 6]
```diff
@@ -208,11 +209,11 @@ class AnalysisManager {
   };
 
   private readonly onAnalysisPoints = ({ analysisId, points }: analysisPoints) => {
-    const handler = this.handlers.get(analysisId);
-    if (handler != null && typeof handler.onAnalysisPoints === "function") {
-      points.forEach(point => ThreadFront.updateMappedLocation(point.frame));
-      handler.onAnalysisPoints(points);
-    }
+    points.forEach(point => {
+      ThreadFront.updateMappedLocation(point.frame);
```
Comment (jcmorrow) [inline packages/protocol/analysisManager.ts:213]: I changed this line. I was a little confused - don't we *always* want to update the frame locations, even if we don't happen to have a handler for this analysis?

File: packages/protocol/graphics.ts [hunk 7]
```diff
@@ -170,50 +64,20 @@ function onMouseEvents(events: MouseEvent[]) {
...
 }
 
-export let hasAllPaintPoints = false;
-
-export const setHasAllPaintPoints = (newValue: boolean) => {
-  hasAllPaintPoints = newValue;
-};
 export const timeIsBeyondKnownPaints = (time: number) =>
-  !hasAllPaintPoints && gPaintPoints[gPaintPoints.length - 1].time < time;
+  fetchingPaints !== Status.Finished && lastReceivedPaintPoint().time < time;
```
Comment (jcmorrow) [inline packages/protocol/graphics.ts:68]: This should move into the PointsCache file.

File: packages/protocol/graphics.ts [hunk 8]
```diff
@@ -170,50 +64,20 @@ function onMouseEvents(events: MouseEvent[]) {
...
-      if (!paintedGraphics) {
-        const { screen, mouse } = await getGraphicsAtTime(ThreadFront.currentTime, false, true);
-        if (screen && mouse) {
-          paintedGraphics = true;
-          paintGraphics(screen, mouse);
-        }
-      }
-    };
-
     const { client } = await import("./socket");
```
Comment (jcmorrow) [inline packages/protocol/graphics.ts:72]: Do we still need `client` here or could we just as easily make this `ReplayClient`? Or should this wait until we move those resources into suspense things?

File: packages/protocol/graphics.ts [hunk 9]
```diff
@@ -506,13 +337,9 @@ export function refreshGraphics() {
   const graphicsVideo = document.getElementById("graphicsVideo") as HTMLVideoElement;
 
   // Find an image to draw.
-  let image;
-  if (gDrawImage) {
+  let image = gLastImage;
```
Comment (jcmorrow) [inline packages/protocol/graphics.ts:340]: Do we still *have* a `gLastImage`?

File: src/ui/components/ProtocolTimeline.tsx [hunk 10]
```diff
@@ -70,13 +79,17 @@ const Spans = ({
...
+  const firstPaint = imperitavelyGetPaintPointForTime(0)!;
+  const lastPaint = lastReceivedPaintPoint();
+  const replayClient = useContext(ReplayClientContext);
 
   if (!showProtocolTimeline || !loadedRegions) {
     return null;
   }
 
+  // This is doing a fetch, which is silly, this should be elsewhere
+  const paint = getPaintPointForTime(replayClient, 1);
```
Comment (jcmorrow) [inline src/ui/components/ProtocolTimeline.tsx:91]: This can be deleted completely now.

File: src/ui/components/ProtocolTimeline.tsx [hunk 11]
```diff
@@ -70,13 +79,17 @@ const Spans = ({
 export default function ProtocolTimeline() {
   const loadedRegions = useAppSelector(getLoadedRegions);
   const { value: showProtocolTimeline } = useFeature("protocolTimeline");
-  const firstPaint = gPaintPoints[0];
-  const lastPaint = gPaintPoints[gPaintPoints.length - 1];
+  const firstPaint = imperitavelyGetPaintPointForTime(0)!;
```
Comment (jcmorrow) [inline src/ui/components/ProtocolTimeline.tsx:82]: I guess we could do like a `getNextPaintAfter` and pass in the last paint we have seen?

**Review by jcmorrow — COMMENTED** [review]
Gonna fix all of these things

File: packages/protocol/analysisManager.ts (hunk 6 above)
Comment (markerikson) [reply packages/protocol/analysisManager.ts:213]: FYI Brian, Holger, and I had a discussion about this last week. We're probably going to _stop_ mutating frame locations shortly - there's only a few places where those locations get used.

File: packages/protocol/graphics.ts (hunk 9 above)
Comment (jcmorrow) [reply packages/protocol/graphics.ts:335]: Yeah, this is some other stuff, ignore this.

File: src/ui/components/ProtocolTimeline.tsx (hunk 11 above)
Comment (jcmorrow) [reply src/ui/components/ProtocolTimeline.tsx:80]: Nah, this is fine as is.

File: packages/protocol/analysisManager.ts (hunk 6 above)
Comment (jcmorrow) [reply packages/protocol/analysisManager.ts:213]: Very good, I'll leave it then! Thanks!

**Comment by jcmorrow** [comment]
This is super old and realistically not that useful anymore. Closing.

**Pull request closed without merging**